File: sdk-typescript/src/governance/remediation-v1.ts
import { sha256Hex } from '../policy/decision-record.js';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import {
  STRICT_IDENTIFIER_MAX_BYTES,
  STRICT_SET_MAX_ITEMS,
  boundedCanonicalText,
  compareCodePoints,
} from './strict-canonical.js';

export const REMEDIATION_PLAN_V1_SCHEMA = 'obsvr-remediation-plan-v1' as const;
export const REMEDIATION_RETRY_V1_SCHEMA = 'obsvr-remediation-retry-v1' as const;
export const REMEDIATION_PLAN_HASH_DOMAIN = 'obsvr-remediation-plan/1' as const;

const HASH_RE = /^[0-9a-f]{64}$/;
const OUTCOMES = new Set(['MODIFY', 'STEP_UP', 'DEFER']);
const REQUIREMENT_KINDS = new Set([
  'approval', 'context', 'modification', 'scope', 'tool', 'verification', 'wait_until',
]);

export interface RemediationRequirementV1Input {
  kind: 'approval' | 'context' | 'modification' | 'scope' | 'tool' | 'verification' | 'wait_until';
  code: string;
  evidence_key: string;
  expected_value_hash?: string;
  guidance?: string;
}

export interface RemediationPlanV1Input {
  plan_id: string;
  attempt_id: string;
  receipt_hash: string;
  outcome: 'MODIFY' | 'STEP_UP' | 'DEFER';
  reason_code: string;
  requirements: RemediationRequirementV1Input[];
  created_at_ms: number;
  expires_at_ms?: number;
}

export interface RemediationPlanV1Document extends RemediationPlanV1Input {
  schema: typeof REMEDIATION_PLAN_V1_SCHEMA;
}

export interface SatisfiedRemediationRequirementV1Input {
  code: string;
  evidence_hash: string;
}

export interface RemediationRetryV1Input {
  retry_attempt_id: string;
  plan: RemediationPlanV1Input | RemediationPlanV1Document;
  satisfied_requirements: SatisfiedRemediationRequirementV1Input[];
}

export interface RemediationRetryV1Document {
  schema: typeof REMEDIATION_RETRY_V1_SCHEMA;
  retry_attempt_id: string;
  parent_attempt_id: string;
  parent_receipt_hash: string;
  remediation_plan_hash: string;
  satisfied_requirements: SatisfiedRemediationRequirementV1Input[];
}

export class RemediationV1ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'RemediationV1ValidationError';
  }
}

function fail(message: string): never {
  throw new RemediationV1ValidationError(message);
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) fail(`${field} must be an object`);
  return value as Record<string, unknown>;
}

function exactKeys(value: Record<string, unknown>, allowed: readonly string[], field: string): void {
  const unknown = Object.keys(value).filter((key) => !allowed.includes(key)).sort(compareCodePoints);
  if (unknown.length > 0) fail(`${field} contains unsupported field: ${unknown[0]}`);
}

function text(value: unknown, field: string): string {
  return boundedCanonicalText(value, field, STRICT_IDENTIFIER_MAX_BYTES, fail);
}

function hash(value: unknown, field: string): string {
  if (typeof value !== 'string' || !HASH_RE.test(value)) {
    fail(`${field} must be exactly 64 lowercase hexadecimal characters`);
  }
  return value;
}

function integer(value: unknown, field: string): number {
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
    fail(`${field} must be a nonnegative safe integer`);
  }
  return value;
}

function requirement(value: unknown, index: number): RemediationRequirementV1Input {
  const field = `requirements[${index}]`;
  const item = record(value, field);
  exactKeys(item, ['kind', 'code', 'evidence_key', 'expected_value_hash', 'guidance'], field);
  if (typeof item.kind !== 'string' || !REQUIREMENT_KINDS.has(item.kind)) {
    fail(`${field}.kind is unsupported`);
  }
  const result: RemediationRequirementV1Input = {
    kind: item.kind as RemediationRequirementV1Input['kind'],
    code: text(item.code, `${field}.code`),
    evidence_key: text(item.evidence_key, `${field}.evidence_key`),
  };
  if (Object.prototype.hasOwnProperty.call(item, 'expected_value_hash')) {
    result.expected_value_hash = hash(item.expected_value_hash, `${field}.expected_value_hash`);
  }
  if (Object.prototype.hasOwnProperty.call(item, 'guidance')) {
    result.guidance = text(item.guidance, `${field}.guidance`);
  }
  return result;
}

export function buildRemediationPlanV1(input: RemediationPlanV1Input): RemediationPlanV1Document {
  const root = record(input, 'remediation plan');
  exactKeys(root, ['schema', 'plan_id', 'attempt_id', 'receipt_hash', 'outcome', 'reason_code',
    'requirements', 'created_at_ms', 'expires_at_ms'], 'remediation plan');
  if (Object.prototype.hasOwnProperty.call(root, 'schema')
    && root.schema !== REMEDIATION_PLAN_V1_SCHEMA) fail('remediation plan schema is invalid');
  if (typeof root.outcome !== 'string' || !OUTCOMES.has(root.outcome)) {
    fail('outcome must be MODIFY, STEP_UP, or DEFER');
  }
  if (!Array.isArray(root.requirements) || root.requirements.length === 0
    || root.requirements.length > STRICT_SET_MAX_ITEMS) {
    fail(`requirements must contain between 1 and ${STRICT_SET_MAX_ITEMS} items`);
  }
  const requirements = root.requirements.map(requirement).sort((left, right) => compareCodePoints(
    `${left.kind}\0${left.code}\0${left.evidence_key}`,
    `${right.kind}\0${right.code}\0${right.evidence_key}`,
  ));
  if (new Set(requirements.map((item) => item.code)).size !== requirements.length) {
    fail('requirement codes must be unique');
  }
  const created = integer(root.created_at_ms, 'created_at_ms');
  const document: RemediationPlanV1Document = {
    schema: REMEDIATION_PLAN_V1_SCHEMA,
    plan_id: text(root.plan_id, 'plan_id'),
    attempt_id: text(root.attempt_id, 'attempt_id'),
    receipt_hash: hash(root.receipt_hash, 'receipt_hash'),
    outcome: root.outcome as RemediationPlanV1Document['outcome'],
    reason_code: text(root.reason_code, 'reason_code'),
    requirements,
    created_at_ms: created,
  };
  if (Object.prototype.hasOwnProperty.call(root, 'expires_at_ms')) {
    const expires = integer(root.expires_at_ms, 'expires_at_ms');
    if (expires <= created) fail('expires_at_ms must be after created_at_ms');
    document.expires_at_ms = expires;
  }
  return document;
}

export function canonicalizeRemediationPlanV1(input: RemediationPlanV1Input): string {
  return canonicalJsonForHash(buildRemediationPlanV1(input));
}

export function remediationPlanV1Hash(input: RemediationPlanV1Input): string {
  return sha256Hex(`${REMEDIATION_PLAN_HASH_DOMAIN}\0${canonicalizeRemediationPlanV1(input)}`);
}

export function buildRemediationRetryV1(input: RemediationRetryV1Input): RemediationRetryV1Document {
  const root = record(input, 'remediation retry');
  exactKeys(root, ['retry_attempt_id', 'plan', 'satisfied_requirements'], 'remediation retry');
  const plan = buildRemediationPlanV1(root.plan as RemediationPlanV1Input);
  const retryAttemptId = text(root.retry_attempt_id, 'retry_attempt_id');
  if (retryAttemptId === plan.attempt_id) fail('retry_attempt_id must identify a new attempt');
  if (!Array.isArray(root.satisfied_requirements)
    || root.satisfied_requirements.length > STRICT_SET_MAX_ITEMS) {
    fail(`satisfied_requirements must contain at most ${STRICT_SET_MAX_ITEMS} items`);
  }
  const satisfied = root.satisfied_requirements.map((candidate, index) => {
    const field = `satisfied_requirements[${index}]`;
    const item = record(candidate, field);
    exactKeys(item, ['code', 'evidence_hash'], field);
    return { code: text(item.code, `${field}.code`), evidence_hash: hash(item.evidence_hash, `${field}.evidence_hash`) };
  }).sort((left, right) => compareCodePoints(left.code, right.code));
  if (new Set(satisfied.map((item) => item.code)).size !== satisfied.length) {
    fail('satisfied requirement codes must be unique');
  }
  const requiredCodes = plan.requirements.map((item) => item.code).sort(compareCodePoints);
  if (canonicalJsonForHash(satisfied.map((item) => item.code)) !== canonicalJsonForHash(requiredCodes)) {
    fail('satisfied_requirements must provide evidence for every plan requirement');
  }
  return {
    schema: REMEDIATION_RETRY_V1_SCHEMA,
    retry_attempt_id: retryAttemptId,
    parent_attempt_id: plan.attempt_id,
    parent_receipt_hash: plan.receipt_hash,
    remediation_plan_hash: remediationPlanV1Hash(plan),
    satisfied_requirements: satisfied,
  };
}

export function remediationRetryV1Hash(input: RemediationRetryV1Input): string {
  return sha256Hex(canonicalJsonForHash(buildRemediationRetryV1(input)));
}

File: sdk-python/obsvr/remediation_v1.py
"""Deterministic remediation requirements and linked retry attempts."""

from __future__ import annotations

import hashlib
from typing import Any, Dict, List

from .strict_canonical import (
    STRICT_IDENTIFIER_MAX_BYTES,
    STRICT_SET_MAX_ITEMS,
    bounded_canonical_text,
    code_point_key,
)
from .tool_pinning import _canonical_json_for_hash

REMEDIATION_PLAN_V1_SCHEMA = "obsvr-remediation-plan-v1"
REMEDIATION_RETRY_V1_SCHEMA = "obsvr-remediation-retry-v1"
REMEDIATION_PLAN_HASH_DOMAIN = b"obsvr-remediation-plan/1"
_HEX = frozenset("0123456789abcdef")
_MAX_SAFE_INTEGER = 9_007_199_254_740_991
_OUTCOMES = frozenset({"MODIFY", "STEP_UP", "DEFER"})
_REQUIREMENT_KINDS = frozenset(
    {"approval", "context", "modification", "scope", "tool", "verification", "wait_until"}
)


class RemediationV1ValidationError(ValueError):
    """The remediation statement is not bounded or internally consistent."""


def _fail(message: str) -> None:
    raise RemediationV1ValidationError(message)


def _record(value: Any, field: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        _fail(f"{field} must be an object")
    return value


def _exact_keys(value: Dict[str, Any], allowed: set[str], field: str) -> None:
    unknown = sorted(set(value) - allowed, key=code_point_key)
    if unknown:
        _fail(f"{field} contains unsupported field: {unknown[0]}")


def _text(value: Any, field: str) -> str:
    return bounded_canonical_text(value, field, STRICT_IDENTIFIER_MAX_BYTES, _fail)


def _hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(character not in _HEX for character in value)
    ):
        _fail(f"{field} must be exactly 64 lowercase hexadecimal characters")
    return value


def _integer(value: Any, field: str) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < 0
        or value > _MAX_SAFE_INTEGER
    ):
        _fail(f"{field} must be a nonnegative safe integer")
    return value


def _requirement(value: Any, index: int) -> Dict[str, Any]:
    field = f"requirements[{index}]"
    item = _record(value, field)
    _exact_keys(
        item,
        {"kind", "code", "evidence_key", "expected_value_hash", "guidance"},
        field,
    )
    if item.get("kind") not in _REQUIREMENT_KINDS:
        _fail(f"{field}.kind is unsupported")
    result = {
        "kind": item["kind"],
        "code": _text(item.get("code"), f"{field}.code"),
        "evidence_key": _text(item.get("evidence_key"), f"{field}.evidence_key"),
    }
    if "expected_value_hash" in item:
        result["expected_value_hash"] = _hash(
            item["expected_value_hash"], f"{field}.expected_value_hash"
        )
    if "guidance" in item:
        result["guidance"] = _text(item["guidance"], f"{field}.guidance")
    return result


def build_remediation_plan_v1(input_value: Dict[str, Any]) -> Dict[str, Any]:
    root = _record(input_value, "remediation plan")
    _exact_keys(
        root,
        {
            "schema",
            "plan_id",
            "attempt_id",
            "receipt_hash",
            "outcome",
            "reason_code",
            "requirements",
            "created_at_ms",
            "expires_at_ms",
        },
        "remediation plan",
    )
    if "schema" in root and root["schema"] != REMEDIATION_PLAN_V1_SCHEMA:
        _fail("remediation plan schema is invalid")
    if root.get("outcome") not in _OUTCOMES:
        _fail("outcome must be MODIFY, STEP_UP, or DEFER")
    raw_requirements = root.get("requirements")
    if (
        not isinstance(raw_requirements, list)
        or not raw_requirements
        or len(raw_requirements) > STRICT_SET_MAX_ITEMS
    ):
        _fail(f"requirements must contain between 1 and {STRICT_SET_MAX_ITEMS} items")
    requirements = sorted(
        (_requirement(item, index) for index, item in enumerate(raw_requirements)),
        key=lambda item: code_point_key(
            f"{item['kind']}\x00{item['code']}\x00{item['evidence_key']}"
        ),
    )
    if len({item["code"] for item in requirements}) != len(requirements):
        _fail("requirement codes must be unique")
    created = _integer(root.get("created_at_ms"), "created_at_ms")
    document = {
        "schema": REMEDIATION_PLAN_V1_SCHEMA,
        "plan_id": _text(root.get("plan_id"), "plan_id"),
        "attempt_id": _text(root.get("attempt_id"), "attempt_id"),
        "receipt_hash": _hash(root.get("receipt_hash"), "receipt_hash"),
        "outcome": root["outcome"],
        "reason_code": _text(root.get("reason_code"), "reason_code"),
        "requirements": requirements,
        "created_at_ms": created,
    }
    if "expires_at_ms" in root:
        expires = _integer(root["expires_at_ms"], "expires_at_ms")
        if expires <= created:
            _fail("expires_at_ms must be after created_at_ms")
        document["expires_at_ms"] = expires
    return document


def canonicalize_remediation_plan_v1(input_value: Dict[str, Any]) -> str:
    return _canonical_json_for_hash(build_remediation_plan_v1(input_value))


def remediation_plan_v1_hash(input_value: Dict[str, Any]) -> str:
    payload = (
        REMEDIATION_PLAN_HASH_DOMAIN
        + b"\x00"
        + canonicalize_remediation_plan_v1(input_value).encode("utf-8")
    )
    return hashlib.sha256(payload).hexdigest()


def build_remediation_retry_v1(input_value: Dict[str, Any]) -> Dict[str, Any]:
    root = _record(input_value, "remediation retry")
    _exact_keys(
        root,
        {"retry_attempt_id", "plan", "satisfied_requirements"},
        "remediation retry",
    )
    plan = build_remediation_plan_v1(root.get("plan"))
    retry_attempt_id = _text(root.get("retry_attempt_id"), "retry_attempt_id")
    if retry_attempt_id == plan["attempt_id"]:
        _fail("retry_attempt_id must identify a new attempt")
    raw_satisfied = root.get("satisfied_requirements")
    if not isinstance(raw_satisfied, list) or len(raw_satisfied) > STRICT_SET_MAX_ITEMS:
        _fail(
            f"satisfied_requirements must contain at most {STRICT_SET_MAX_ITEMS} items"
        )
    satisfied: List[Dict[str, str]] = []
    for index, candidate in enumerate(raw_satisfied):
        field = f"satisfied_requirements[{index}]"
        item = _record(candidate, field)
        _exact_keys(item, {"code", "evidence_hash"}, field)
        satisfied.append(
            {
                "code": _text(item.get("code"), f"{field}.code"),
                "evidence_hash": _hash(
                    item.get("evidence_hash"), f"{field}.evidence_hash"
                ),
            }
        )
    satisfied.sort(key=lambda item: code_point_key(item["code"]))
    if len({item["code"] for item in satisfied}) != len(satisfied):
        _fail("satisfied requirement codes must be unique")
    required_codes = sorted(
        (item["code"] for item in plan["requirements"]), key=code_point_key
    )
    if [item["code"] for item in satisfied] != required_codes:
        _fail("satisfied_requirements must provide evidence for every plan requirement")
    return {
        "schema": REMEDIATION_RETRY_V1_SCHEMA,
        "retry_attempt_id": retry_attempt_id,
        "parent_attempt_id": plan["attempt_id"],
        "parent_receipt_hash": plan["receipt_hash"],
        "remediation_plan_hash": remediation_plan_v1_hash(plan),
        "satisfied_requirements": satisfied,
    }


def remediation_retry_v1_hash(input_value: Dict[str, Any]) -> str:
    return hashlib.sha256(
        _canonical_json_for_hash(build_remediation_retry_v1(input_value)).encode("utf-8")
    ).hexdigest()


__all__ = [
    "REMEDIATION_PLAN_V1_SCHEMA",
    "REMEDIATION_RETRY_V1_SCHEMA",
    "RemediationV1ValidationError",
    "build_remediation_plan_v1",
    "build_remediation_retry_v1",
    "canonicalize_remediation_plan_v1",
    "remediation_plan_v1_hash",
    "remediation_retry_v1_hash",
]
